### PR TITLE
Support built-in id function for TensorVariable on parameters

### LINKED ISSUE #130087: Calling id function on model param causes error or graph break for compile
### 🐛 Describe the bug

When compiling a model, if built-in id function on the parameter Tensor is used in model code, it will cause an error if full_graph=True and it will cause graph breaks for full_graph=False. We saw this problem in DeepSpeed Stage 1 and 2 optimizer when running DeepSpeed with compiled autograd.

The following is the example demonstrate this problem (Of course, DeepSpeed is using similar logic in optimizer code instead model forward code):

```
import torch

class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        # constant tensors (similar to the model parameters such as weights)
        self.y1 = torch.ones(2)
        self.y2 = torch.zeros(2)
        
        self.ref_y1_id = id(self.y1)
        self.ref_y2_id = id(self.y2)
    
    def forward(self, x, ref_id):
        if ref_id == id(self.y1):
            x = torch.mul(x, self.y1)
        else:
            x = torch.mul(x, self.y2)
        return x


x = torch.ones(2)

m1 = M()
compiled_m1 = torch.compile(m1, fullgraph=True)

prediction = compiled_m1(x, m1.ref_y1_id)
prediction = compiled_m1(x, m1.ref_y2_id)
```
The above code will show the following error:

```
  File "/home/haifchen/working/envs/env_tensor_id/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 1520, in call_id
    unimplemented(f"call_id with args {args}")
  File "/home/haifchen/working/envs/env_tensor_id/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 190, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_id with args (TensorVariable(),)

from user code:
   File "/home/haifchen/working/test/test-tensor-id.py", line 94, in forward
    if ref_id == id(self.y1):
```

The basic thing it is doing is it call id function on param and use the resulted id to identify a param. Although we can use directly the self.ref_y1_id to replace  id(self.y1) easily in the above code. While for many optimizers that was written, the calling path and code is quite complex and to refactor the code to use the id value initialized at the beginning is not a trivial work.

It is necessary to support built-in id function call on parameters (tensors) in such use cases, so that user model code or optimizer code doesn't need to change to support compile mode or compiled autograd without graph breaks.

If it can be done, it is a great user experience improvement for enabling torch compile or compiled autograd.

### Versions

Collecting environment information...
PyTorch version: 2.3.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-107-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 1
Core(s) per socket:                 6
Socket(s):                          2
Stepping:                           0
BogoMIPS:                           4190.15
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xsaves arat pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
Hypervisor vendor:                  VMware
Virtualization type:                full
L1d cache:                          384 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           12 MiB (12 instances)
L3 cache:                           55 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX flush not necessary, SMT disabled
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Syscall hardening, KVM SW loop
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] torch==2.3.1
[pip3] triton==2.3.1
[conda] Could not collect

cc @ezyang @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames

### PR DESCRIPTION
**Pull request by jerrychenhf**
Fixes #130087

This patch tries to provide a built-in id function implementation for TensorVariable when the id function is called on tensors like module parameters. The id function call on intermediate tensors is not supported.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -414,6 +414,24 @@ def try_generic_attr_handling():
             raise NotImplementedError
         return result
 
+    def call_id(self, tx):
+        if not self.source:
+            raise NotImplementedError
```
Comment (anijain2305) [inline torch/_dynamo/variables/tensor.py:419]: Use `unimplemented` to cause a graph break.

Comment (anijain2305) [reply torch/_dynamo/variables/tensor.py:419]: Same applies to all the time you are using using `NotImplementedError` below.

**Review by anijain2305 — CHANGES_REQUESTED** [review]
Needs a small change of using `unimplemented`. Else LGTM.

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (jerrychenhf) [reply torch/_dynamo/variables/tensor.py]: Done

**Comment by jerrychenhf** [comment]
@anijain2305 Thank you for reviewing!

Done replacing NotImplementedError with unimplemented.

**Review by anijain2305 — APPROVED** [review]

**Comment by jerrychenhf** [comment]
The failures are related to the new added test_inline_inbuilt_nn_modules.py which runs with _dynamo.config.inline_inbuilt_nn_modules=True. This gives different op_count with or without this condition for the test_id_tensor case.
I will try to remove the op_count assert for such cases.

**Comment by jerrychenhf** [comment]
@anijain2305 Thank you for approving.
The latest PR should fix the failures from test_inline_inbuilt_nn_modules.

**Review by anijain2305 — APPROVED** [review]

**Comment by jerrychenhf** [comment]
Seems the remaining failures (inductor/test_minifier.py::MinifierTests::test_accuracy_vs_strict_accuracy) is unrelated to this PR.

**Comment by jerrychenhf** [comment]
@anijain2305 Thank you for reviewing!
Would you please help a check whether I need to anything before this can be merged?

**Comment by anijain2305** [comment]
Can you please rebase once? I can take it from there.

**Pull request closed without merging**